### ISSUE TITLE
chore(common): add new PR control commands to manage builds

### ISSUE THREAD

**Issue by mcdurdin** [keymanapp/keyman#8361]
It would be useful to be able to skip certain builds or do only builds without artifacts etc. Rationale: many of our minor changes are triggering many builds, and while this is a good default, we can reduce the burden on the build agents by being a little more circumspect when we know certain platforms will not be impacted. We can also reduce the artifact burden and build times by skipping release builds when they are not needed.

These commands would override the defaults -- so you could force a build where it would not automatically run otherwise, for example. The following commands would be supported:

* `skip`: Skip build altogether for a given target:

  ```
  Build-bot: skip windows,developer
  ```

* `build`: Just compile and run unit tests on given target:

  ```
  Build-bot: build windows,developer
  ```

* `release`: Make a full installer for a given target and make it available as an artifact on the PR, good for user testing:

  ```
  Build-bot: release windows,developer
  ```

**QUESTION:** Do we need to support additional commands or parameters for commands? For example, Developer has a set of low-cost unit tests, plus a very long regression test suite (25+ minutes), which we could do with `test(quick)`, `test(full)` and the default (`test`) would be `test(quick)`. Something like that?

  - use `test-quick` vs `test-full` vs `test` perhaps, and we may be able to pass that into build.sh in the future as well.
  - PR builds default to quick tests (but full installer builds?) -- but can override
  - nightly builds do full tests
  - master-merge builds do not build installers

Multiple commands in a comment should be supported.

The following build targets should be defined:
* `android`
* `common` (for tests)
* `core` (for tests)
* `developer`
* `ios`
* `linux`
* `mac`
* `web`
* `windows`

**QUESTION:** Am I missing any? Perhaps the ability to specify platform for common + core would be helpful?

  - ability to skip build of sample apps would help

Multiple targets should be comma separated, suggest no spaces between to support future parameters.

The commands could be:
* `Build-bot: skip [target[,...]]` -- don't build anything for these target(s)
* `Build-bot: test [target[,...]]` -- only compile and run automated tests for these target(s)
* `Build-bot: release [target[,...]]` -- `test` + build an installable artifact for these target(s), if applicable (i.e. `release common` would not be an error but would only run tests for `common`)

These commands will be checked on all comments in a PR and are consolidated; later commands override earlier commands.

**IDEA:** reduce set of default builds, e.g. don't build Developer,Android,iOS if Web builds?

**Comment by mcdurdin** [comment]
Developer build time, rounded, things that could be eliminated during a shorter test build:

Time (min) | Process
-----------|----------
26:30 | regression tests against keyboards repo (additional flag required?)
4:30 | build installer
0:45 | signcode
0:15 | design-time packages
2:30 | excess `npm ci` calls
1:15 | upload artifacts
**36:45** | **total skippable steps**
**47:45** | **total build time**
**11:00** | **new build time!**

**Comment by darcywong00** [comment]
> The commands could be:
>
>no -- don't build anything for this target

I was hoping we could reuse `skip` like with the test-bot. Otherwise, I can easily see my future self mixing up no/skip in my PRs...

**Comment by mcdurdin** [comment]
Suggest doing this as a commit message trailer instead of as a PR command. Ties the implementation to looking at the last commit which avoids a bunch of potential conflicts.

`Keyman-build-bot: <command>`

**Comment by mcdurdin** [comment]
See also #13212

**Comment by mcdurdin** [comment]
iOS deployment to TestFlight is a good trial balloon for this

**Comment by ermshiperete** [comment]
> Suggest doing this as a commit message trailer instead of as a PR command. Ties the implementation to looking at the last commit which avoids a bunch of potential conflicts.
> 
> `Keyman-build-bot: <command>`

I'm not convinced. Then we have to remember to re-add the trailer if we add another commit, e.g. after addressing code review comments. And it wouldn't work when we commit suggestions on GH.

Unless the bot would ignore later commits without `Keyman-build-bot` trailer and continue to use the last one it found in the PR commits (or the defaults if there are none in the commits of the PR).

**Comment by mcdurdin** [comment]
> Unless the bot would ignore later commits without Keyman-build-bot trailer and continue to use the last one it found in the PR commits (or the defaults if there are none in the commits of the PR).

Yes, the bot would consolidate all commands found. I am considering checking this _only_ in commits for consistency -- because that avoids confusion about when a the command is checked, as a build would not be re-run if you add a comment later in the PR.

(For performance and reliability, we would look up the PR commit history to determine which commits belong in a given PR. As a messed-up PR can pull in wrong commits, we'd have a threshold e.g. 100 commits, after which, only the default builds would run.)

Alternative: only run builds if a Build-bot: PR comment is found. But risky because just editing a PR comment could cause an expensive re-run of builds. So, my pref is currently to look only at commit messages. (Force-push an amended commit would be acceptable way to trigger builds if needed.)

**Comment by mcdurdin** [comment]
Completed in #14125
